### ISSUE TITLE
Building large data structures eats up way too much memory

### ISSUE THREAD

**Issue by LourensVeen** [multiscale/muscle3#75]
Building a list of 10M items each consisting of a list containing a 3-item list, a double and a small integer takes tens of GB of memory, while the actual amount required for the MsgPack-encoded data is only about 390MB. It looks like the internal data structures are really inefficient memory-wise.

The work-around for now is to use a list or dict of Grids, as grids are more efficient than lists.

Digging into this a bit, the problem is that we have a lot of zones, and that those zones are _very_ expensive. Each item is a Data at some point, and gets its own messagepack::object in its own zone, and those zones are all collected in the vector of zones of the final Data object. We have 10M list items, with 7 items each (outer list, inner list, 3 doubles inside it, another double, and an int), so we have 70M zones and objects.

A messagepack::object is 16 bytes, and each zone has a pointer to it in DCR and uses a minimum of 64 bytes for its internal accounting. So we get 70M * 88 = ~= 6GB of overhead at least. That's nothing compared to the memory allocated within the zones themselves though, which change size as necessary, but start out at 8kB! So 70M zones, times 8kB is 560GB! Ouch!

Setting the zones to be very small by default should help a lot. If we use 16 bytes each then we reduce that 560GB to 1.1GB, which is more manageable. It's still three times the original data size though, and there's the other 6GB of overhead as well.

A better solution is to copy data across into the list's zone when assigning an object to a list item, then the original zone can be deallocated. For the above example, that should reduce the overhead to 50M non-list objects times 16 bytes = 800MB, plus 20M lists times 16 + 64 bytes each = 1.6 GB. There will still be a lot of allocating and deallocating going on, so it's not too efficient.

Another improvement may be to make `mp_obj_` a direct member of DCR, rather than a pointer to an object on a zone. That means the zone can remain at 0 bytes allocated unless this DCR represents a dict or a list and we need to allocate items. If we also do lazy allocation of the zone itself in DCR, then a DCR containing a double will use 16 bytes for `mp_obj_` and 16 bytes for an empty `mp_zones_`, which is nice especially if it's a temporary and we'll end up just with just 32 bytes on the stack and copying 8 bytes in and out of it.

**Comment by LourensVeen** [comment]
Urgh, that commit is unrelated, it should have mentioned #74.

**Comment by LourensVeen** [comment]
Okay, I've made some improvements which should get the original 10M-object problem down to 5.4GB, which is still a lot but also more than a factor 100 better than what we had. I've also added a note to the documentation to recommend the use of grids if you have large amounts of data, which is the real solution.

It's possible to optimise irregular data more, but we're approaching the point where I end up designing my own data representation and data structures, and while I would love to have a go, I don't have the time.
